flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimeServiceManager.java

```diff
@@ -74,19 +74,16 @@
 
 	private final Map<String, InternalTimerServiceImpl<K, ?>> timerServices;
 
-	private final boolean useLegacySynchronousSnapshots;
-
 	InternalTimeServiceManager(
-		KeyGroupRange localKeyGroupRange,
-		KeyContext keyContext,
-		PriorityQueueSetFactory priorityQueueSetFactory,
-		ProcessingTimeService processingTimeService, boolean useLegacySynchronousSnapshots) {
```

Comment:
Why is `useLegacySynchronousSnapshots` removed here? Can't `RocksDBKeyedStateBackend.requiresLegacySynchronousTimerSnapshots()` still return true here?

Reply:
It can. Let me explain it though, why I think it is unnecessary in the ctor. (There is one thing I forgot in the PR, I wanted to make the `snapshotStateForKeyGroup` private. It is used only in the `InternalTimeServiceManager#snapshotState`.)

The flag is used only in the `snapshotStateForKeyGroup` which is used only in the `snapshotState` method. The `snapshotState` method looks like:
```
	public void snapshotState(
			KeyedStateBackend<?> keyedStateBackend,
			StateSnapshotContext context,
			String operatorName) throws Exception {
		//TODO all of this can be removed once heap-based timers are integrated with RocksDB incremental snapshots
		if (keyedStateBackend instanceof AbstractKeyedStateBackend &&
			((AbstractKeyedStateBackend<?>) keyedStateBackend).requiresLegacySynchronousTimerSnapshots()) {

            ....
			snapshotStateForKeyGroup(new DataOutputViewStreamWrapper(out), keyGroupIdx);
            ....
		}
	}
```

Effectively this flag is checked twice. Once in the preconditions against the value in ctor and the second time in the `snapshotState` method against the value set in the `KeyedStateBackend`. In my opinion the one time in the `snapshotState` is enough.

The only case when the condition in `snapshotStateForKeyGroup` could fail is if we created the `InternalTimeServiceManager` from a different `KeyedStateBackend` than we use for snapshotting. I am quite sure this would let to way more problems than only this precondition failling.

Reply:
It seems after the changes made for FLINK-16316 the `useLegacySynchronousSnapshots` field here is never accessed? I'm also +1 for keeping the private field and use it in `snapshotState` to save the cost of `instanceof` when calling `requiresLegacySynchronousTimerSnapshots` for RocksDB backend.

Reply:
I think I changed it the way you suggested @carp84